staging/src/k8s.io/legacy-cloud-providers/azure/azure.go

```diff
@@ -242,21 +273,70 @@ func NewCloud(configReader io.Reader) (cloudprovider.Interface, error) {
 		config.VMType = vmTypeStandard
 	}
 
+	if config.CloudConfigType == "" {
+		// The default cloud config type is cloudConfigTypeMerge.
+		config.CloudConfigType = cloudConfigTypeMerge
+	} else {
+		supportedCloudConfigTypes := sets.NewString(
+			string(cloudConfigTypeMerge),
+			string(cloudConfigTypeFile),
+			string(cloudConfigTypeSecret))
+		if !supportedCloudConfigTypes.Has(string(config.CloudConfigType)) {
+			return fmt.Errorf("cloudConfigType %v is not supported, supported values are %v", config.CloudConfigType, supportedCloudConfigTypes.List())
+		}
+	}
+
+	if config.CloudConfigScope == "" {
+		// The default config scope is cloudConfigScopeAll.
+		config.CloudConfigScope = cloudConfigScopeAll
+	} else {
+		supportedCloudConfigScopes := sets.NewString(
+			string(cloudConfigScopeAll),
+			string(cloudConfigScopeNode),
+			string(cloudConfigScopeControlPlane))
+		if !supportedCloudConfigScopes.Has(string(config.CloudConfigScope)) {
+			return fmt.Errorf("cloudConfigScope %v is not supported, supported values are %v", config.CloudConfigScope, supportedCloudConfigScopes.List())
+		}
+	}
+
 	env, err := auth.ParseAzureEnvironment(config.Cloud)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
 	servicePrincipalToken, err := auth.GetServicePrincipalToken(&config.AzureAuthConfig, env)
 	if err == auth.ErrorNoAuth {
-		if !config.UseInstanceMetadata {
-			// No credentials provided, useInstanceMetadata should be enabled.
-			return nil, fmt.Errorf("useInstanceMetadata must be enabled without Azure credentials")
+		runingAsKubelet, err := isRunningAsKubelet()
```

Comment:
Same comment as earlier, a kubelet check in the cloud provider seems awkward to me and backwards to out-of-tree patterns.

Reply:
won't remove this requirement.